sbom.py

```diff
@@ -0,0 +1,306 @@
+        "name": "CPython",
+        "versionInfo": cpython_version,
+        "licenseConcluded": "PSF-2.0",
+        "originator": "Organization: Python Software Foundation",
+        "supplier": "Organization: Python Software Foundation",
+        "packageFileName": tarball_name,
+        "externalRefs": [
+            {
+                "referenceCategory": "SECURITY",
+                "referenceLocator": f"cpe:2.3:a:python:python:{cpython_version}:*:*:*:*:*:*:*",
+                "referenceType": "cpe23Type",
+            }
+        ],
+        "primaryPackagePurpose": "SOURCE",
+        "downloadLocation": tarball_download_location,
+        "checksums": [{"algorithm": "SHA256", "checksumValue": tarball_checksum_sha256}],
+    }
+
+    # The top-level CPython package depends on every vendored sub-package.
+    for sbom_package in sbom["packages"]:
+        sbom["relationships"].append({
+            "spdxElementId": sbom_cpython_package["SPDXID"],
+            "relatedSpdxElement": sbom_package["SPDXID"],
+            "relationshipType": "DEPENDS_ON",
+        })
+
+    sbom["packages"].append(sbom_cpython_package)
+
+    # Extract all currently known files from the SBOM with their checksums.
+    known_sbom_files = {}
+    for sbom_file in sbom["files"]:
+        sbom_filename = sbom_file["fileName"]
+
+        # Look for the expected SHA256 checksum.
+        for sbom_file_checksum in sbom_file["checksums"]:
+            if sbom_file_checksum["algorithm"] == "SHA256":
+                known_sbom_files[sbom_filename] = (
+                    sbom_file_checksum["checksumValue"]
+                )
+                break
+        else:
+            raise ValueError(
+                f"Couldn't find expected SHA256 checksum in SBOM for file '{sbom_filename}'"
+            )
+
+    # Now we walk the tarball and compare known files to our expected checksums in the SBOM.
+    # All files that aren't already in the SBOM can be added as "CPython" files.
+    for member in tarball.getmembers():
+        if member.isdir():  # Skip directories!
+            continue
+
+        # Get the member from the tarball. CPython prefixes all of its
+        # source code with 'Python-{version}/...'.
+        assert member.isfile() and member.name.startswith(f"Python-{cpython_version}/")
+
+        # Calculate the hashes, either for comparison with a known value
+        # or to embed in the SBOM as a new file. SHA1 is only used because
+        # SPDX requires it for all file entries.
+        file_bytes = tarball.extractfile(member).read()
+        actual_file_checksum_sha1 = hashlib.sha1(file_bytes).hexdigest()
+        actual_file_checksum_sha256 = hashlib.sha256(file_bytes).hexdigest()
+
+        # Remove the 'Python-{version}/...' prefix for the SPDXID and fileName.
+        member_name_no_prefix = member.name.split('/', 1)[1]
+
+        # We've already seen this file, so we check it hasn't been modified and continue on.
```

Comment:
Tar files can contain the same file multiple times, there isn't a situation where this can lead to false positives here, is there? (I'm thinking in terms of the listed sha256 being the same but other information being different in the tarball-contained SBOM.) I guess we're trusting both the tarball and the contained SBOM anyway, so no.